app/src/main/java/com/spascoding/taskycourse/feature_auth/presentation/login/LoginViewModel.kt

```diff
@@ -29,32 +32,78 @@ class LoginViewModel @Inject constructor(
 
     fun onEvent(event: LoginEvent) {
         when (event) {
-            is LoginEvent.ChangeEmail -> {
+            is LoginEvent.ChangeEmail -> { handleChangeEmailAction(event) }
+            is LoginEvent.ChangePassword -> { handleChangePasswordAction(event) }
+            is LoginEvent.LoginAction -> { handleLoginAction(event) }
+            is LoginEvent.SignUpAction -> {}
+        }
+    }
+
+    private fun handleChangeEmailAction(event: LoginEvent.ChangeEmail) {
+        UserDataValidator.validateEmail(event.email)
+            .onSuccess {
                 state.update {
-                    it.copy(email = event.email)
+                    it.copy(
+                        email = event.email,
+                        validEmail = true,
+                    )
                 }
             }
-            is LoginEvent.ChangePassword -> {
+            .onError {
                 state.update {
-                    it.copy(password = event.password)
+                    it.copy(
+                        email = event.email,
+                        validEmail = false,
+                    )
                 }
             }
-            is LoginEvent.LoginAction -> {
-                viewModelScope.launch(Dispatchers.IO) {
-                    authRepository.login(
-                        email = state.value.email,
-                        password = state.value.password,
-                    ).onError { error ->
-                        val errorMassage = error.asUiText()
-                        _toastChannel.send(UserEvent.Error(errorMassage))
-                    }
+    }
+    private fun handleChangePasswordAction(event: LoginEvent.ChangePassword) {
+        UserDataValidator.validateEmail(event.password)
+            .onSuccess {
+                state.update {
+                    it.copy(
+                        password = event.password,
+                        validPassword = true,
+                    )
                 }
             }
-            is LoginEvent.SignUpAction -> {}
+            .onError {
+                state.update {
+                    it.copy(
+                        password = event.password,
+                        validPassword = false,
+                    )
+                }
+            }
+    }
+    private fun handleLoginAction(event: LoginEvent.LoginAction) {
+        viewModelScope.launch(Dispatchers.IO) {
+            UserDataValidator.validateEmail(event.email)
+                .onError { error ->
+                    val errorMassage = error.asUiText()
+                    _toastChannel.send(UserEvent.Error(errorMassage))
+                    return@launch
+                }
+
+            UserDataValidator.validatePassword(event.password)
+                .onError { error ->
+                    val errorMassage = error.asUiText()
+                    _toastChannel.send(UserEvent.Error(errorMassage))
+                    return@launch
+                }
```

Comment:
This will only validate the inputs one at a time. Wouldn't it be better usability to show all errors at once?